### PR TITLE
Sticky Scroll : when several outline models are available, choose the one spanning the biggest total range

### LINKED ISSUE #159136: repeated line in sticky scroll
Does this issue occur when all extensions are disabled?: Yes/No

版本: 1.71.0-insider (user setup)
提交: 7ed02a69658fc07f08740b98fa0d8d96c1a61ed4
日期: 2022-08-24T05:18:03.269Z
Electron: 19.0.12
Chromium: 102.0.5005.167
Node.js: 16.14.2
V8: 10.2.154.15-electron.0
OS: Windows_NT x64 10.0.25182
沙盒化: Yes

Steps to Reproduce:

image

### PR DESCRIPTION
**Pull request by aiday-mar**
In particular it fixes https://github.com/microsoft/vscode/issues/159136

sticky-scroll-multiple-providers

To test:

Open a launch.json file
Enable sticky scroll and verify that the sticky lines stuck to the top of the view port do not correspond to the outline model in the bottom-right pannel

### CODE REVIEW

**Comment by heartacker** [comment]
Why do we merge this in next release?
August is unreleased

**Comment by aiday-mar** [comment]
@heartacker the PR has not yet been thoroughly tested and reviewed so we leave some time to be sure it behaves correctly and does not break anything else. This change will be placed into next month's stable release. You can however view the new modifications on Insiders once the PR is merged.

File: src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts [hunk 1]
```diff
@@ -84,14 +91,44 @@ export class StickyLineCandidateProvider extends Disposable {
...
+						const totalSumRanges = this.findSumOfRangesOfGroup(outlineGroup);
+						if (totalSumRanges > maxTotalSumOfRanges) {
+							maxTotalSumOfRanges = totalSumRanges;
+							providerString = key;
+						}
+					}
+					this._providerString = providerString;
+					outlineModel = outlineModel.children.get(this._providerString) as unknown as OutlineModel;
+				}
+			}
```
Comment (jrieken) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts:131]: Move into `fromOutlineModel`

File: src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts [hunk 2]
```diff
@@ -84,14 +91,44 @@ export class StickyLineCandidateProvider extends Disposable {
...
 			const model = this._editor.getModel();
 			const modelVersionId = model.getVersionId();
-			const outlineModel = await OutlineModel.create(this._languageFeaturesService.documentSymbolProvider, model, token) as OutlineModel;
+			let outlineModel = await OutlineModel.create(this._languageFeaturesService.documentSymbolProvider, model, token) as OutlineModel;
 			if (token.isCancellationRequested) {
 				return;
 			}
+			// When several possible outline providers
+			if (outlineModel.children.size !== 0 && outlineModel.children.values().next().value instanceof OutlineGroup) {
+				if (outlineModel.children.has(this._providerString)) {
```
Comment (jrieken) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts:116]: 💄 `const [firstChild] = outlineModel.children.values();` or `Iterable.first`

File: src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts [hunk 3]
```diff
@@ -84,14 +92,44 @@ export class StickyLineCandidateProvider extends Disposable {
...
+					let maxTotalSumOfRanges = 0;
+					for (const [key, outlineGroup] of outlineModel.children.entries()) {
+						const totalSumRanges = this.findSumOfRangesOfGroup(outlineGroup);
+						if (totalSumRanges > maxTotalSumOfRanges) {
+							maxTotalSumOfRanges = totalSumRanges;
+							providerString = key;
+						}
+					}
+					this._providerString = providerString;
+					outlineModel = outlineModel.children.get(this._providerString) as unknown as OutlineModel;
```
Comment (jrieken) [inline src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts:130]: 👮🏼‍♂️ `as unknown as Something` is not allowed and a hint that types are wrong. Either spec the `outlineModel` as or-type or use dedicated variables

**Review by jrieken — CHANGES_REQUESTED** [review]

File: src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts (hunk 3 above)
Comment (aiday-mar) [reply src/vs/editor/contrib/stickyScroll/browser/stickyScrollProvider.ts]: Yes that commit was just an initial commit. I committed the new changes.

**Review by jrieken — DISMISSED** [review]

**Review by jrieken — APPROVED** [review]

**Pull request merged**